src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java

```diff
@@ -1461,4 +1467,29 @@ private boolean onSpongeLightningDamage(net.minecraft.entity.Entity entity, Dama
         }
     }
 
+    @Inject(method = "setDead", at = @At(value = "RETURN"))
+    private void onSetDead(CallbackInfo ci) {
+        if (ShouldFire.DESTRUCT_ENTITY_EVENT && !this.world.isRemote && !(EntityUtil.toNative(this) instanceof EntityLiving)) {
+            MessageChannel originalChannel = MessageChannel.TO_NONE;
+            try (final CauseStackManager.StackFrame frame = Sponge.getCauseStackManager().pushCauseFrame()) {
+                Object source = PhaseTracker.getInstance().getCurrentContext().getSource();
+                frame.pushCause(source);
```

Comment:
I'm almost sure that at this point the source will already be pushed to the frame stack from the PhaseStates populating lazily the cause stack. Might want to check if that is indeed the case or not here.